File: enterprise-dotnet-api/src/AILearnAPI.Application/Services/InterviewRoadmapService.cs
using AILearnAPI.Application.Interfaces;
using AILearnAPI.Domain.Entities;
using AILearnAPI.Domain.Interfaces;
using AILearnAPI.Shared.DTOs.InterviewRoadmap;

namespace AILearnAPI.Application.Services
{
    public class InterviewRoadmapService : IInterviewRoadmapService
    {
        private readonly IInterviewRoadmapRepository _repo;

        public InterviewRoadmapService(IInterviewRoadmapRepository repo)
        {
            _repo = repo;
        }

        // ── Queries ──────────────────────────────────────────────────────────

        public async Task<List<InterviewRoadmapDto>> GetByUserIdAsync(string userId)
        {
            var roadmaps = await _repo.GetByUserIdAsync(userId);
            return roadmaps.Select(ToDto).ToList();
        }

        public async Task<InterviewRoadmapDto?> GetByUserAndStackAsync(string userId, string techStackId)
        {
            var rm = await _repo.GetByUserAndStackAsync(userId, techStackId.ToLowerInvariant());
            return rm == null ? null : ToDto(rm);
        }

        // ── Commands ─────────────────────────────────────────────────────────

        public async Task<InterviewRoadmapDto> SaveAsync(string userId, SaveInterviewRoadmapDto dto)
        {
            // Try to load existing to preserve createdAt
            var existing = await _repo.GetByUserAndStackAsync(userId, dto.techStackId.ToLowerInvariant());

            var roadmap = existing ?? new InterviewRoadmap
            {
                UserId       = userId,
                TechStackId  = dto.techStackId.ToLowerInvariant(),
            };

            roadmap.TechStackName  = dto.techStackName.Trim();
            roadmap.TechStackIcon  = dto.techStackIcon.Trim();
            roadmap.LastAccessedAt = DateTime.UtcNow;
            roadmap.UpdatedAt      = DateTime.UtcNow;
            roadmap.Sections       = dto.sections.Select(MapSection).ToList();

            var saved = await _repo.UpsertAsync(roadmap);
            return ToDto(saved);
        }

        public async Task<InterviewRoadmapDto?> UpdateProgressAsync(
            string userId, string roadmapId, UpdateProgressDto dto)
        {
            var roadmap = await _repo.GetByIdAsync(roadmapId);
            if (roadmap == null || roadmap.UserId != userId) return null;

            // Build quick lookup: topicId → done
            var lookup = dto.topics.ToDictionary(t => t.id, t => t.done);

            foreach (var sec in roadmap.Sections)
            {
                foreach (var topic in sec.Topics)
                {
                    if (lookup.TryGetValue(topic.Id, out var done))
                    {
                        if (done && !topic.Done)
                            topic.CompletedAt = DateTime.UtcNow;
                        else if (!done)
                            topic.CompletedAt = null;

                        topic.Done = done;
                    }
                }
            }

            roadmap.UpdatedAt      = DateTime.UtcNow;
            roadmap.LastAccessedAt = DateTime.UtcNow;

            var updated = await _repo.UpdateAsync(roadmapId, roadmap);
            return updated == null ? null : ToDto(updated);
        }

        public async Task<bool> DeleteAsync(string userId, string roadmapId)
        {
            return await _repo.DeleteByUserAndIdAsync(userId, roadmapId);
        }

        // ── Mapping ──────────────────────────────────────────────────────────

        private static InterviewRoadmapDto ToDto(InterviewRoadmap r)
        {
            var sectionDtos = r.Sections.Select(s => new RoadmapSectionDto
            {
                id       = s.Id,
                title    = s.Title,
                emoji    = s.Emoji,
                expanded = s.Expanded,
                topics   = s.Topics.Select(t => new RoadmapTopicDto
                {
                    id          = t.Id,
                    text        = t.Text,
                    done        = t.Done,
                    completedAt = t.CompletedAt
                }).ToList()
            }).ToList();

            var allTopics = sectionDtos.SelectMany(s => s.topics).ToList();
            var done  = allTopics.Count(t => t.done);
            var total = allTopics.Count;

            return new InterviewRoadmapDto
            {
                id             = r.Id,
                userId         = r.UserId,
                techStackId    = r.TechStackId,
                techStackName  = r.TechStackName,
                techStackIcon  = r.TechStackIcon,
                sections       = sectionDtos,
                doneCount      = done,
                totalCount     = total,
                percent        = total == 0 ? 0 : (int)Math.Round((double)done / total * 100),
                createdAt      = r.CreatedAt,
                updatedAt      = r.UpdatedAt,
                lastAccessedAt = r.LastAccessedAt
            };
        }

        private static RoadmapSectionData MapSection(RoadmapSectionDto s) => new()
        {
            Id       = s.id,
            Title    = s.title,
            Emoji    = s.emoji,
            Expanded = s.expanded,
            Topics   = s.topics.Select(t => new RoadmapTopicData
            {
                Id   = t.id,
                Text = t.text,
                Done = t.done,
                CompletedAt = t.completedAt
            }).ToList()
        };
    }
}

File: angular-starter/src/app/services/interview-roadmap.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { CustomAuthService } from '../shared/custom-auth.service';
import { RoadmapSection } from '../interview-prep/interview-prep.component';

// ─── Backend DTOs (mirror of C# Shared DTOs) ─────────────────────────────────

export interface RoadmapTopicDto {
  id:          string;
  text:        string;
  done:        boolean;
  completedAt: string | null;
}

export interface RoadmapSectionDto {
  id:       string;
  title:    string;
  emoji:    string;
  expanded: boolean;
  topics:   RoadmapTopicDto[];
}

export interface InterviewRoadmapDto {
  id:             string;
  userId:         string;
  techStackId:    string;
  techStackName:  string;
  techStackIcon:  string;
  sections:       RoadmapSectionDto[];
  doneCount:      number;
  totalCount:     number;
  percent:        number;
  createdAt:      string;
  updatedAt:      string;
  lastAccessedAt: string;
}

export interface SaveRoadmapPayload {
  techStackId:   string;
  techStackName: string;
  techStackIcon: string;
  sections:      RoadmapSectionDto[];
}

export interface UpdateProgressPayload {
  topics: Array<{ id: string; done: boolean }>;
}

// ─────────────────────────────────────────────────────────────────────────────

@Injectable({ providedIn: 'root' })
export class InterviewRoadmapService {

  private readonly url = '/api/interview-roadmap';

  constructor(
    private http:    HttpClient,
    private authSvc: CustomAuthService,
  ) {}

  // ── Read ──────────────────────────────────────────────────────────────────

  /** Load all saved roadmaps for the current user. Returns [] on error. */
  getAll(): Observable<InterviewRoadmapDto[]> {
    return this.http.get<InterviewRoadmapDto[]>(this.url, {
      headers: this.authSvc.getAuthHeaders()
    }).pipe(catchError(() => of([])));
  }

  /** Load the roadmap for a specific tech-stack. Returns null if not saved yet. */
  getByStack(techStackId: string): Observable<InterviewRoadmapDto | null> {
    return this.http.get<InterviewRoadmapDto>(
      `${this.url}/stack/${techStackId}`,
      { headers: this.authSvc.getAuthHeaders() }
    ).pipe(catchError(() => of(null)));
  }

  // ── Write ─────────────────────────────────────────────────────────────────

  /** Upsert a full roadmap (structure + initial done state). */
  save(payload: SaveRoadmapPayload): Observable<InterviewRoadmapDto | null> {
    return this.http.post<InterviewRoadmapDto>(this.url, payload, {
      headers: this.authSvc.getAuthHeaders()
    }).pipe(catchError(() => of(null)));
  }

  /** Patch only topic done-states for an already-saved roadmap. */
  updateProgress(roadmapId: string, sections: RoadmapSection[]): Observable<InterviewRoadmapDto | null> {
    const topics: Array<{ id: string; done: boolean }> = [];
    sections.forEach(sec => sec.topics.forEach(t => topics.push({ id: t.id, done: t.done })));
    const payload: UpdateProgressPayload = { topics };
    return this.http.put<InterviewRoadmapDto>(
      `${this.url}/${roadmapId}/progress`,
      payload,
      { headers: this.authSvc.getAuthHeaders() }
    ).pipe(catchError(() => of(null)));
  }

  delete(roadmapId: string): Observable<boolean> {
    return this.http.delete(`${this.url}/${roadmapId}`, {
      headers: this.authSvc.getAuthHeaders()
    }).pipe(
      map(() => true),
      catchError(() => of(false))
    );
  }

  // ── Helpers ───────────────────────────────────────────────────────────────

  /** Convert backend sections → RoadmapSection[] (for rehydrating the component). */
  toRoadmapSections(dto: InterviewRoadmapDto): RoadmapSection[] {
    return dto.sections.map((s, i) => ({
      id:       s.id,
      title:    s.title,
      emoji:    s.emoji,
      expanded: i === 0, // keep first section open
      topics:   s.topics.map(t => ({
        id:   t.id,
        text: t.text,
        done: t.done,
      })),
    }));
  }
}
